### PR TITLE
Create SRP library

### PR DESCRIPTION
**Pull request by Gymbotan**
We want to realize SRP protocol that used in Proton realization. And create Client that can communicate with proton email server.
In that PR created srp library (server, client, utils, additional tools) based on BounceCastle solution. Added two tests.

### CODE REVIEW

File: TuviSRPLib/ProtonBase64Lib/ProtonBase64.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+using System.Numerics;
+using System.Text;
+
+namespace ProtonBase64Lib
+{
+    public static class ProtonBase64
+    {
+        private const int BaseValue = 64;
+        private const int ByteSize = 8;
+        private const int SixBitsSize = 6;
```
Comment (anatolse) [inline TuviSRPLib/ProtonBase64Lib/ProtonBase64.cs:11]: I would choose another name, probably with an explanation that we converting from base 2^8 to base 2^6, for example, the target unit size is equal to log2(64), something like that

File: TuviSRPLib/TuviSRPLib/ProtonSRPUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+using Org.BouncyCastle.Crypto;
+using Org.BouncyCastle.Crypto.Generators;
+using Org.BouncyCastle.Math;
+using Org.BouncyCastle.Security;
+using Org.BouncyCastle.Utilities;
+using ProtonBase64Lib;
+
+namespace TuviSRPLib
+{
+    public class ProtonSRPUtilities
```
Comment (anatolse) [inline TuviSRPLib/TuviSRPLib/ProtonSRPUtilities.cs:13]: Let's make this class `static`

File: TuviSRPLib/TuviSRPLib/ProtonSRPUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        private const string proton = "proton";
+        private const int cost = 10;
```
Comment (anatolse) [inline TuviSRPLib/TuviSRPLib/ProtonSRPUtilities.cs:16]: Please, add a comment. What does this `cost` mean?

File: TuviSRPLib/TuviSRPLib/ProtonSRPClient.cs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+using Org.BouncyCastle.Security;
+
+namespace TuviSRPLib
+{
+    /**
+	 * Implements the client side SRP-6a protocol. Note that this class is stateful, and therefore NOT threadsafe.
+	 * This implementation of SRP is based on the optimized message sequence put forth by Thomas Wu in the paper
+	 * "SRP-6: Improvements and Refinements to the Secure Remote Password Protocol, 2002"
+	 */
+    public class ProtonSRPClient
```
Comment (anatolse) [inline TuviSRPLib/TuviSRPLib/ProtonSRPClient.cs:14]: Please, add a comment that this is an excerpt from BounsyCastle lib

File: TuviSRPLib/TuviSRPLib/ProtonSRPServer.cs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+using Org.BouncyCastle.Security;
+
+namespace TuviSRPLib
+{
+    /**
+	 * Implements the server side SRP-6a protocol. Note that this class is stateful, and therefore NOT threadsafe.
+	 * This implementation of SRP is based on the optimized message sequence put forth by Thomas Wu in the paper
+	 * "SRP-6: Improvements and Refinements to the Secure Remote Password Protocol, 2002"
+	 */
+    public class ProtonSRPServer
```
Comment (anatolse) [inline TuviSRPLib/TuviSRPLib/ProtonSRPServer.cs:14]: Please, add a comment that this is an excerpt from BounsyCastle lib
Add the link to the original code

File: TuviSRPLib/TuviSRPLibTests/ProtonBase64Tests.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
+﻿using ProtonBase64Lib;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TuviSRPLibTests
+{
+    internal class ProtonBase64Tests
```
Comment (anatolse) [inline TuviSRPLib/TuviSRPLibTests/ProtonBase64Tests.cs:10]: Please, add tests with incorrect arguments

File: TuviSRPLib/ProtonBase64Lib/BigIntegerExtension.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+using System.Collections.Generic;
+using System.Numerics;
+using System.Text;
+
+namespace ProtonBase64Lib
+{
+    /// <summary>
+    /// Extra function to work with BigInteger.
+    /// </summary>
+    public static class BigIntegerExtension
```
Comment (anatolse) [inline TuviSRPLib/ProtonBase64Lib/BigIntegerExtension.cs:11]: Please, add tests for this extension

File: TuviSRPLib/TuviSRPLibTests/ProtonSRPTests.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        [Test]
+        public void FullCycleOfWorkTest()
+        {
+            var encodedN = "W2z5HBi8RvsfYzZTS7qBaUxxPhsfHJFZpu3Kd6s1JafNrCCH9rfvPLrfuqocxWPgWDH2R8neK7PkNvjxto9TStuY5z7jAzWRvFWN9cQhAKkdWgy0JY6ywVn22+HFpF4cYesHrqFIKUPDMSSIlWjBVmEJZ/MusD44ZT29xcPrOqeZvwtCffKtGAIjLYPZIEbZKnDM1Dm3q2K/xS5h+xdhjnndhsrkwm9U9oyA2wxzSXFL+pdfj2fOdRwuR5nW0J2NFrq3kJjkRmpO/Genq1UW+TEknIWAb6VzJJJA244K/H8cnSx2+nSNZO3bbo6Ys228ruV9A8m6DhxmS+bihN3ttQ==";
+            var decodedN = Base64.Decode(encodedN);
+            BigInteger N = new BigInteger(1, decodedN.Reverse().ToArray());
+            BigInteger g = new BigInteger("2");
+
+            string identity = "ivanov";
```
Comment (anatolse) [inline TuviSRPLib/TuviSRPLibTests/ProtonSRPTests.cs:19]: replace with culture-independent value, for example, "testidentity"

File: TuviSRPLib/TuviSRPLibTests/ProtonSRPTests.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            Assert.IsTrue(client.VerifyServerEvidenceMessage(M2), "Message M2 is not verified.");
+
+            BigInteger clientKey = client.CalculateSessionKey();
+            BigInteger serverKey = server.CalculateSessionKey();
+
+            Assert.AreEqual(clientKey, serverKey);
+        }
+
+        [Test]
+        public void ClientSideWorkTest()
```
Comment (anatolse) [inline TuviSRPLib/TuviSRPLibTests/ProtonSRPTests.cs:57]: I would added a negative test, just to be sure that authentication fails if the client provides incorrect data

File: TuviSRPLib/ProtonBase64Lib/ProtonBase64.cs [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    /// </summary>
+    public static class ProtonBase64
+    {
+        /// <summary>
+        /// Current Base64 algorithm tramsforms bytes(8-bits) sequence into 6-bits sequence. 
+        /// Then converts 6-bits elements into symbols of choosen dictionary.
+        /// So, to calculate size of this sequences we need following parameters:
+        /// </summary>
+        private const int BaseValue = 64;
+        private const int EightBitsSize = 8; // byte size
```
Comment (anatolse) [inline TuviSRPLib/ProtonBase64Lib/ProtonBase64.cs:18]: I would use `sourceElementBitSize` and `targetElementBitSize`

**Review by ValeraFinebits — APPROVED** [review]

**Review by anatolse — APPROVED** [review]

**Pull request merged**